core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java

```diff
@@ -290,110 +271,112 @@ private boolean checkCacheNotFoundResponseInPartitionHandling(TransactionBoundar
       return false;
    }
 
-   private boolean shouldFetchRemoteValuesForWriteSkewCheck(InvocationContext ctx, WriteCommand cmd) {
-      // Note: the primary owner always already has the data, so this method is always going to return false
-      if (useClusteredWriteSkewCheck && ctx.isInTxScope() && dm.isRehashInProgress()) {
-         for (Object key : cmd.getAffectedKeys()) {
-            boolean shouldPerformWriteSkewCheck = cdl.localNodeIsPrimaryOwner(key);
-            // TODO Dan: remoteGet() already checks if the key is available locally or not
-            if (shouldPerformWriteSkewCheck && dm.isAffectedByRehash(key) && !dataContainer.containsKey(key))
-               return true;
-         }
-      }
-      return false;
-   }
-
    /**
     * If we are within one transaction we won't do any replication as replication would only be performed at commit
     * time. If the operation didn't originate locally we won't do any replication either.
     */
-   private Object handleTxWriteCommand(InvocationContext ctx, WriteCommand command, Object key, boolean skipRemoteGet) throws Throwable {
+   private Object handleTxWriteCommand(InvocationContext ctx, WriteCommand command, Object key) throws Throwable {
       // see if we need to load values from remote sources first
-      if (!skipRemoteGet && needValuesFromPreviousOwners(ctx, command))
-         remoteGetBeforeWrite(ctx, command, key);
+      remoteGetBeforeWrite(ctx, command, key);
 
-      // FIRST pass this call up the chain.  Only if it succeeds (no exceptions) locally do we attempt to distribute.
       return invokeNextInterceptor(ctx, command);
    }
 
    @Override
-   protected boolean needValuesFromPreviousOwners(InvocationContext ctx, WriteCommand command) {
+   protected boolean writeNeedsRemoteValue(InvocationContext ctx, WriteCommand command, Object key) {
+      if (command.hasFlag(Flag.CACHE_MODE_LOCAL)) {
+         return false;
+      }
       if (ctx.isOriginLocal()) {
          // The return value only matters on the originator.
          // Conditional commands also check the previous value only on the originator.
-         if (isNeedReliableReturnValues(command) || command.isConditional())
-            return true;
+         if (!command.readsExistingValues()) {
+            return false;
+         }
+         // TODO Could make DELTA_WRITE/ApplyDeltaCommand override SKIP_REMOTE_LOOKUP by changing next line to
+         // return !command.hasFlag(Flag.SKIP_REMOTE_LOOKUP) || command.alwaysReadsExistingValues();
+         if (command.hasFlag(Flag.SKIP_REMOTE_LOOKUP)) {
```

Comment:
It seems this could be changed to ``return !command.hasFlag(Flag.SKIP_REMOTE_LOOKUP))``